### PR TITLE
Problem: Hare doesn't handle node failure properly

### PR DESCRIPTION
**Pull request by mssawant**
If a node is rebooted Consul does report a warning for the corresponding
service to hax but hax checks for node status which is 'passing' atleast
for few iterations, this delay in node status update leads to missing of
the Consul warning, Consul status for the process is not update in KV.
Consul KV for the process status still remains M0_CONF_HA_PROCESS_STARTED
which also causes node-join to fail.
While notifying service failure to peer motr processes it is also important
to handle delay in notification mainly in case of transient failure.
The process for which Consul reported failure might just be starting and
reporting a process failure when hax has already replied to the entrypoint
request leads to other inconsistencies with halink protocol.
Hax also does not send failure notifications for the disks corresponding
to an ioservice. This is needed to select the appropriate pool version
when writing post node failure.

Solution:
- When Consul reports process failure before broadcasting the failure to
  motr processes, check the service status one more time before broadcasting.
  This takes care of delay in failure notification if the process has already
  started. If the process has already restarted hax has already notified peer motr
  processes about this motr process on receiving its first entrypoint request.
- Update Consul status when broadcasting failure, this handles node rejoining.
- Send failure notifications for the disks corressponding to a failed ioservice.
- increase max size of consumer queue to 512, noticed a hang in processing
  in case of heavy message flow.

### CODE REVIEW

File: hax/hax/util.py [hunk 1]
```diff
@@ -635,6 +696,10 @@ def ensure_motr_all_started(self, event: Event):
                 return
             wait_for_event(event, 5)
 
+    def fid2objT(self, fid: str) -> str:
```
Comment (knekrasov) [inline hax/hax/util.py:699]: What strings can this method return? If we know the full vocabulary of possible return values, does it make sense to return a enum or something? That would make the contract clear.

File: hax/hax/handler.py [hunk 2]
```diff
@@ -61,23 +61,70 @@ def stop(self) -> None:
...
+        svc_status = m0HaProcessEvent.event_to_svchealth(event.chp_event)
+        if event.chp_type == m0HaProcessType.M0_CONF_HA_PROCESS_M0D:
+            # Broadcast the received motr process status to other motr
+            # processes in the cluster.
+            q.put(BroadcastHAStates(states=[HAState(fid=event.fid,
+                                    status=svc_status)],
+                  reply_to=None))
 
-    def update_process_failure(self, ha_states: List[HAState]) -> None:
+    def update_process_failure(self, q: Queue,
```
Comment (knekrasov) [inline hax/hax/handler.py:79]: It seems like the method needs to be decorated with `@repeat_if_fails()`
Reason: the code within can raise `HAConsistencyException`. The method is invoked directly from `qconsumer` thread (handler.py) and it seems like the exception is not handled there. So `qconsumer` will pull the message from the thread Queue and once the exception raises, the message will be forgotten and will not be processed afterwards.

File: hax/hax/handler.py [hunk 3]
```diff
@@ -61,23 +61,70 @@ def stop(self) -> None:
...
+                    proc_status = self.consul.get_process_status(state.fid)
+                    svc_status = self.consul.get_service_health(node,
+                                                                state.fid.key)
+                    LOG.info('node: %s proc: %s proc_state: %s health: %s',
+                             node, state.fid, proc_status, svc_status)
+                    if svc_status == ServiceHealth.FAILED:
+                        m0status = m0HaProcessEvent.M0_CONF_HA_PROCESS_STOPPED
+                        if proc_status != 'M0_CONF_HA_PROCESS_STOPPED':
+                            pevent = ConfHaProcess(chp_event=m0status,
+                                                   chp_type=3,
```
Comment (knekrasov) [inline hax/hax/handler.py:97]: what `chp_type=3` stands for?

File: hax/hax/hax.py [hunk 4]
```diff
@@ -89,7 +89,7 @@ def main():
     # _run_qconsumer_thread function.
     #
     # [KN] Note: The server is launched in the main thread.
-    q: Queue = Queue(maxsize=8)
+    q: Queue = Queue(maxsize=512)
```
Comment (knekrasov) [inline hax/hax/hax.py:92]: 👍

File: hax/hax/types.py [hunk 5]
```diff
@@ -42,6 +42,16 @@ class Uint128Struct(c.Structure):
 ObjT.__doc__ = 'Motr conf object types and their m0_fid.f_container values'
 
 
+FidT = Enum(
```
Comment (knekrasov) [inline hax/hax/types.py:45]: According to the way how FidT is used, it feels more like a static map `fid container` -> `string name`. Why wouldn't we convert it into the static dict then?

**Review by knekrasov — CHANGES_REQUESTED** [review]

File: hax/hax/handler.py [hunk 6]
```diff
@@ -61,23 +61,70 @@ def stop(self) -> None:
...
         for state in ha_states:
-            if state.status == ServiceHealth.FAILED:
-                m0status = m0HaProcessEvent.M0_CONF_HA_PROCESS_STOPPED
-                pevent = ConfHaProcess(chp_event=m0status,
-                                       chp_type=3,
-                                       chp_pid=0,
-                                       fid=state.fid)
-                self._update_process_status(pevent)
+            if self.consul.fid2objT(str(state.fid)) == 'PROCESS':
+                svc_status = state.status
```
[the remark this answers is not in the document]
Comment (mssawant) [reply hax/hax/handler.py:84]: Putting everything in a single function was a bit tricky so split the code into 2 utility functions.

File: hax/hax/handler.py [hunk 7]
```diff
@@ -61,23 +61,70 @@ def stop(self) -> None:
...
+                    # check one more time before broadcasting failure.
+                    node = self.consul.get_process_node(state.fid)
+                    proc_status = self.consul.get_process_status(state.fid)
+                    svc_status = self.consul.get_service_health(node,
+                                                                state.fid.key)
+                    LOG.info('node: %s proc: %s proc_state: %s health: %s',
+                             node, state.fid, proc_status, svc_status)
+                    if svc_status == ServiceHealth.FAILED:
+                        m0status = m0HaProcessEvent.M0_CONF_HA_PROCESS_STOPPED
+                        if proc_status != 'M0_CONF_HA_PROCESS_STOPPED':
```
[the remark this answers is not in the document]
Comment (mssawant) [reply hax/hax/handler.py:95]: Yeah, removed `proc_status != STOPPED` as it was redundant.

File: hax/hax/handler.py (hunk 3 above)
Comment (mssawant) [reply hax/hax/handler.py:97]: Its `m0HaProcessType.M0_CONF_HA_PROCESS_M0D`, updated.

File: hax/hax/types.py (hunk 5 above)
Comment (mssawant) [reply hax/hax/types.py:45]: Yeah, it was a leftover. fixed the usage.

File: hax/hax/util.py [hunk 8]
```diff
@@ -581,31 +609,64 @@ def set_m0_disk_state(self, fid: str, objstate: int) -> None:
...
         except (ConsulException, HTTPError, RequestException) as e:
             raise HAConsistencyException('Failed to communicate '
                                          'to Consul Agent') from e
         return status
 
+    def get_process_node(self, proc_fid: Fid) -> str:
+        fidk = proc_fid.key
+        # 'node/<node_name>/process/<process_fidk>/service/type'
+        node_items = self.kv.kv_get('m0conf/nodes', recurse=True)
+        # TODO [KN] This code is too cryptic. To be refactored.
```
[the remark this answers is not in the document]
Comment (mssawant) [reply hax/hax/util.py:661]: Removed.

File: hax/hax/util.py (hunk 1 above)
Comment (mssawant) [reply hax/hax/util.py:699]: ObjT for a given fid, added a docstring.

File: hax/hax/util.py [hunk 9]
```diff
@@ -581,31 +609,64 @@ def set_m0_disk_state(self, fid: str, objstate: int) -> None:
         self.kv.kv_put(key, data)
 
     @repeat_if_fails()
-    def get_service_health(self, service_name: str,
-                           node: str, svc_id: int) -> ServiceHealth:
-
+    def get_service_health(self, node: str, svc_id: int) -> ServiceHealth:
```
[the remark this answers is not in the document]
Comment (mssawant) [reply hax/hax/util.py:612]: Added a docstring.

File: hax/hax/util.py [hunk 10]
```diff
@@ -635,6 +696,10 @@ def ensure_motr_all_started(self, event: Event):
                 return
             wait_for_event(event, 5)
 
+    def fid2objT(self, fid: str) -> str:
+        cont = f'0x{Fid.parse("0x7200000000000001:0x9").container:x}'
```
[the remark this answers is not in the document]
Comment (mssawant) [reply hax/hax/util.py]: fixed this.

File: hax/hax/handler.py [hunk 11]
```diff
@@ -4,7 +4,7 @@
 # you may not use this file except in compliance with the License.
 # You may obtain a copy of the License at
 #
-#    http://www.apache.org/licenses/LICENSE-2.0
+#    http://www.apache.org/lic://github.com/Seagate/cortx-motr/blob/b22c3dd234b7885b6528e8870192f00816d75048/pool/pool.c#L659enses/LICENSE-2.0
```
Comment (knekrasov) [inline hax/hax/handler.py:7]: Is this hyperlink correct?

File: hax/hax/util.py (hunk 1 above)
Comment (knekrasov) [reply hax/hax/util.py:703]: My point was that raw strings as result type is error prone and tools like mypy will fail to help if I mistype the string when comparing the result value to something.

But OK, we can address this later in another PR. Would you please add FIXME for the future?

File: hax/hax/types.py (hunk 5 above)
Comment (knekrasov) [reply hax/hax/types.py:45]: Sorry, I still don't understand why you need it to be a Enum. According to what I see, you need the mapping, it doesn't need to be a separate type. And you use it like a dict, actually

```python
    def fid2objT(self, fid: str) -> str:
        """
        Returns ObjT for a given fid as a string.
        """
        cont = f'0x{Fid.parse(fid).container:x}'
        return str(FidT[cont].value)
```

So effectively FidT could either be a dict or a function that performs this mapping (while dict could be an implementation detail). And I would vote for function. Do I miss anything?

**Review by knekrasov — CHANGES_REQUESTED** [review]
A couple of minor things left.

File: hax/hax/handler.py (hunk 11 above)
Comment (suvratjoshi) [reply hax/hax/handler.py:7]: Seems like motr's pool.c link mistakenly got copied here

Comment (mssawant) [reply hax/hax/handler.py:7]: Damn, must have got pasted mistakenly from the cache, restored.

File: hax/hax/util.py (hunk 1 above)
Comment (mssawant) [reply hax/hax/util.py]: No, I agree. I see, this is not needed actually. Have updated the patch.

File: hax/hax/types.py (hunk 5 above)
Comment (mssawant) [reply hax/hax/types.py]: Removed, not needed. Thanks.

**Review by knekrasov — APPROVED** [review]

**Review by suvratjoshi — APPROVED** [review]

**Pull request merged**